source/extensions/omni.isaac.lab/test/controllers/test_operational_space.py

```diff
@@ -0,0 +1,864 @@
+
+    def _update_states(
+        self,
+        robot: Articulation,
+        ee_frame_idx: int,
+        arm_joint_ids: list[int],
+    ):
+        # obtain dynamics related quantities from simulation
+        ee_jacobi_idx = ee_frame_idx - 1
+        jacobian_w = robot.root_physx_view.get_jacobians()[:, ee_jacobi_idx, :, arm_joint_ids]
+        mass_matrix = robot.root_physx_view.get_mass_matrices()[:, arm_joint_ids, :][:, :, arm_joint_ids]
+        gravity = robot.root_physx_view.get_generalized_gravity_forces()[:, arm_joint_ids]
+        # Convert the Jacobian from world to root frame
+        jacobian_b = jacobian_w.clone()
+        root_rot_matrix = matrix_from_quat(quat_inv(robot.data.root_state_w[:, 3:7]))
+        jacobian_b[:, :3, :] = torch.bmm(root_rot_matrix, jacobian_b[:, :3, :])
+        jacobian_b[:, 3:, :] = torch.bmm(root_rot_matrix, jacobian_b[:, 3:, :])
+
+        # Compute current pose of the end-effector
+        root_pose_w = robot.data.root_state_w[:, 0:7]
+        ee_pose_w = robot.data.body_state_w[:, ee_frame_idx, 0:7]
+        ee_pos_b, ee_quat_b = subtract_frame_transforms(
+            root_pose_w[:, 0:3], root_pose_w[:, 3:7], ee_pose_w[:, 0:3], ee_pose_w[:, 3:7]
+        )
+        ee_pose_b = torch.cat([ee_pos_b, ee_quat_b], dim=-1)
+
+        # Compute the current velocity of the end-effector
+        ee_vel_w = robot.data.body_vel_w[:, ee_frame_idx, :]  # Extract end-effector velocity in the world frame
+        root_vel_w = robot.data.root_vel_w  # Extract root velocity in the world frame
+        relative_vel_w = ee_vel_w - root_vel_w  # Compute the relative velocity in the world frame
+        ee_lin_vel_b = quat_rotate_inverse(robot.data.root_quat_w, relative_vel_w[:, 0:3])  # From world to root frame
+        ee_ang_vel_b = quat_rotate_inverse(robot.data.root_quat_w, relative_vel_w[:, 3:6])
+        ee_vel_b = torch.cat([ee_lin_vel_b, ee_ang_vel_b], dim=-1)
+
+        # Calculate the contact force
+        ee_force_w = torch.zeros(self.num_envs, 3, device=self.sim.device)
+        if self.contact_forces is not None:  # Only modify if it exist
+            sim_dt = self.sim.get_physics_dt()
+            self.contact_forces.update(sim_dt)  # update contact sensor
+            # Calculate the contact force by averaging over last four time steps (i.e., to smoothen) and
+            # taking the max of three surfaces as only one should be the contact of interest
+            ee_force_w, _ = torch.max(torch.mean(self.contact_forces.data.net_forces_w_history, dim=1), dim=1)
+
+        # This is a simplification, only for the sake of testing.
+        ee_force_b = ee_force_w
+
+        return jacobian_b, mass_matrix, gravity, ee_pose_b, ee_vel_b, root_pose_w, ee_pose_w, ee_force_b
+
+    def _update_target(
+        self,
+        osc: OperationalSpaceController,
+        root_pose_w: torch.tensor,
+        ee_pose_b: torch.tensor,
+        target_set: torch.tensor,
+        current_goal_idx: int,
+    ):
```

Comment:
please add docstrings for this function

Reply:
"""Update the target for the operational space controller.

        Args:
            osc (OperationalSpaceController): The operational space controller.
            root_pose_w (torch.tensor): The root pose in the world frame.
            ee_pose_b (torch.tensor): The end-effector pose in the body frame.
            target_set (torch.tensor): The target set to track.
            current_goal_idx (int): The current goal index.

        Returns:
            command (torch.tensor): The target command.
            ee_target_pose_b (torch.tensor): The end-effector target pose in the body frame.
            ee_target_pose_w (torch.tensor): The end-effector target pose in the world frame.
            next_goal_idx (int): The next goal index.

        Raises:
            ValueError: If the target type is undefined.
        """